ROFO-74 회원탈퇴 API 작성

Unagi-zoso opened the pull request:
### 개요
회원탈퇴 기능을 작성합니다.
제거 범위는 유저동의약관과 유저, S3 이미지, redis refresh rotate Id 입니다.

### 변경사항
- ImageService remove 메소드 및 테스트 작성
- 유저동의약관 제거 쿼리 메소드 및 테스트 작성 (deleteAllByUser)
- AuthEventHandler 및 AuthLeaveEvent 객체 작성
- 회원탈퇴 API 및 테스트 작성

### 테스트
- 테스트 코드 추가여부 O

### 관련 지라 및 위키 링크
 - [JIRA](https://weit-2nd.atlassian.net/browse/ROFO-74) 


### 리뷰어에게 하고 싶은 말
이벤트 방식으로 구현되어 서비스단 테스트는 통합테스트로만 작성해봤습니다.
이벤트 방식으로 구현은 나름 해봤는데 이런식으로 하는게 맞는지 잘 모르겠네요..
회원탈퇴에 관련한 의존을 Handler 에서 맡은건 좋아보이지만 좀 어설픈것 같아 피드백을 듣고싶습니다.


스웨거 API 에러 부분을 어노테이션으로 대체해볼까 했지만 제가 반환하는 예외 중엔 
동적으로 예외메시지를 만드는 경우가 있어 모든 경우를 어노테이션으로 대체하기는 불가능해 보였습니다.
로컬 브랜치에 작업은 해뒀는데 저만의 기준이 세워지면 따로 PR 을 올리겠습니다.

Unagi-zoso committed:
Merge branch 'develop' into feature/ROFO-74

Unagi-zoso committed:
ROFO-74 Image 제거 기능 작성

Unagi-zoso committed:
ROFO-74 유저동의약관 제거 쿼리 메소드 작성

Unagi-zoso committed:
ROFO-47 AuthEventHandler 및 회원탈퇴 이벤트 객체 작성

Unagi-zoso committed:
ROFO-47 회원탈퇴 메소드 및 테스트 작성

Unagi-zoso committed:
ROFO-47 회원탈퇴 API 작성

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/application/event/AuthEventHandler.kt:
```diff
@@ -0,0 +1,29 @@
+    private val jwtUtil: JwtUtil,
+) {
+    @EventListener(AuthLeaveEvent::class)
+    fun handleAuthLeaveEvent(event: AuthLeaveEvent) {
```
제기억에는 이벤트 핸들러에 @ Async 어노테이션 안붙히면 이벤트 발행에서부터 처리까지 동기로 이루어 지는걸로 알고있는데
그걸 의도하신걸까요??

Unagi-zoso replied:
네. 제 생각에 회원탈퇴를 하고 성공 응답까지 받았는데
이미 발급 받은 accessToken 으로 접근이 가능할 수도 있다 판단했습니다.
이러한 부분을 고려했을때 동기적으로 작동하는게 맞지 않을까 생각을 했습니다.

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/application/service/AuthCommandService.kt:
```diff
@@ -90,4 +93,9 @@ class AuthCommandService(
         jwtUtil.removeCachedRefreshToken(user.id)
     }
+
+    @Transactional
+    fun leave(user: User) {
+        applicationEventPublisher.publishEvent(AuthLeaveEvent(user.id))
+    }
```
오호 ㅋㅋㅋ 여기에 트랜잭셔널을 걸어두신걸보면
이벤트를 동기로 처리하는걸 의도하신게 맞는것 같은데 그러면 이벤트 기반으로 작성한 이유가 조금 퇴색되지 않을까 싶은데
의도를 들어보고 싶습니다!

Unagi-zoso replied:
네. 제가 작업하면서도 마음에 걸려했던 부분이네요.
그래도 회원탈퇴라는 기능은 동기적으로 동작하는게 맞지 않을까 하는 생각이 있었습니다.
혹시 비동기적으로 동작했을 때 문제되는 경우가 없을까요? (User 정보가 완전히 제거되기전 AccessToken 을 사용해 접근한다던가 하는.. 진짜 나쁜 마음 먹은 경우 아니면 안 일어날것 같은데 좀 걱정되어 결정했습니다. )

konempty replied:
네넵 동기로 하는건 상관없는데 그러면 이벤트 퍼블리셔가 아니라 그냥 서비스에 메소드로 만들어서
그 메소드를 사용하도록 하는게 낫지않을까 싶어서 리뷰드렸어요!

Unagi-zoso replied:
네. 말씀해주신대로 그냥 서비스에 바로 구현해버리는 것과 차이가 없는게 아닌가 생각이 드네요.
근데 또 회원탈퇴를 위해선 회원과 연관된 부분들을 제거하기 위해 그에 따른 의존이 생길텐데
회원탈퇴만을 위해 앞으로 생길 의존을 서비스에서 다 주입 받는건 조금 마음에 걸리는 부분이 있네요.
그리고 회원탈퇴 기능이 현 시스템 상에서 정말 동기적으로만 동작해야되는지에 대해서도 아직 고민이 되서
도메인 담당 시에 언급해주셨던 것 처럼 비동기 전환을 염두해두고 있습니다.

따라서 회원탈퇴만을 위한 의존을 서비스에서 분리할 수 있다는 장점과 이후에 있을 비동기 방식으로의 전환을 고려했을 때
현재 동기식 이벤트의 형태를 유지해도 될까요?

(비동기 기능을 바로 적용시키고 싶지만 고려해야할 요소들이 꽤 많아서 
다음 작업을 회원탈퇴 기능의 비동기 방식 전환으로 해도 괜찮을까요?)

konempty replied:
> 회원탈퇴만을 위해 앞으로 생길 의존을 서비스에서 다 주입 받는건 조금 마음에 걸리는 부분이 있네요.

아 ㅋㅋㅋ 맞네요 요건 잘못된게 맞는것 같아요
그렇다면 지금처럼 이벤트 기반으로 두는게 더 예쁠것 같습니다~!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/presentation/spec/AuthControllerSpec.kt:
```diff
@@ -348,4 +348,51 @@ interface AuthControllerSpec {
+                description = "회원탈퇴 성공",
+            ),
+            ApiResponse(
+                responseCode = "401",
+                description = "유효하지 않은 토큰으로 요청",
+                content = [
+                    Content(
+                        mediaType = MediaType.APPLICATION_JSON_VALUE,
+                        schema = Schema(implementation = ErrorResponse::class),
+                        examples = [
+                            ExampleObject(
+                                name = "Unauthenticated Access",
+                                summary = "인증되지 않은 접근",
+                                value = """
+                        {
+                            "code": -10001,
+                            "errorMessage": "인증정보가 없습니다."
+                        }
+                        """,
+                            ),
+                            ExampleObject(
+                                name = "Authenticated User Not Found",
+                                summary = "존재하지 않는 사용자",
+                                value = """
+                        {
+                            "code": -10001,
+                            "errorMessage": "인증된 사용자를 찾을 수 없습니다."
+                        }
+                        """,
+                            ),
+                        ],
+                    ),
+                ],
+            ),
```
요 에러부분은 지언님이 작업해주신 방향으로 하는게 더 깔끔할것 같습니다!

Unagi-zoso replied:
네. 다른 분들 작업보니 확실히 좋더라구요.
하지만 일단 이번 작업에 적용하는건 좀 꺼렸습니다.
제가 만든 예외 중엔 동적으로 예외메시지를 만드는 경우가 있어 ErrorCode 에 정적으로 만들기가
애매했습니다. 작업을 한다면 채린님이 작업해주신 것 처럼 동적인 예외는 현재의 방식을 고수하고
정적으로 나타낼 수 있는 것은 ErrorCode 에 명시해 어노테이션 방식으로 깔끔히 처리할 수 있을 듯 합니다.
현재 PR 은 회원탈퇴 API 를 위한 것이라 이왕 어노테이션 방식을 적용한다면 새로 PR 을 만들어 적용하는게 어떨까 생각했습니다.

그냥 현재 PR 에 추가해도 괜찮을까요?

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/presentation/spec/AuthControllerSpec.kt:
```diff
@@ -348,4 +348,51 @@ interface AuthControllerSpec {
+        ],
+    )
+    fun withdraw(
+        @Parameter(hidden = true)
+        user: User,
```
일일히 유저를 숨기지말고
스프링독 설정에서 글로벌하게 ignore 처리를 해두시는게 좋을것 같습니다!

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/application/service/AuthIntegrationServiceTest.kt:
```diff
@@ -211,4 +212,19 @@ class AuthIntegrationServiceTest(
+                    val user = userRepository.getByNickname(signUpRequest.nickname)
+                    authCommandService.leave(user)
+                    shouldThrow<UserNotFoundException> { userRepository.getByNickname(signUpRequest.nickname) }
+                    val isStored = redisTemplate.hasKey(getRefreshTokenCacheKey(user.id))
+                    isStored.shouldBeFalse()
```
유저가 로그인한적이 없으니까 유저 탈퇴이전에도
isStored.shouldBeFalse()가 성립하지않나요??
그렇다면 특정 행동 전후의 변화검증이 아닐것 같아요

Unagi-zoso replied:
아하. 제가 안드로이드 개발자 분께 의견을 받아 회원가입 시에
별도의 로그인 과정 없이도 accessToken, refreshToken 을 바로 발급할 수 있게 회원가입 로직을
변경했습니다. 그래서 이 부분을 테스트에서 검증하고자 추가했습니다.

konempty replied:
아하 ㅋㅋㅋ 이해했습니다!

konempty reviewed (commented):
크 너무 고생많으셨습니다 👍 👍 
바로 어프루브해도 될정도의 리뷰양인것 같긴한데
몇가지 진짜 궁금한게 있어서 일단 코멘트로 남겨두겠습니다 ㅋㅋㅋ

Unagi-zoso committed:
ROFO-74 Swagger User 파라미터 전역 범위 숨김 적용

Unagi-zoso committed:
ROFO-74 Auth API 커스텀 예외 및 ErrorCode 작성

Unagi-zoso committed:
ROFO-74 addExamplesToResponses 로직 수정

기존 방식은 Swagger 제공 @Operation 의 example 에 있는 내용을 덮어씌워 기존 내용이 유실됩니다.
기존 내용을 포함하도록 변경했습니다.

Unagi-zoso committed:
ROFO-74 API 문서 ErrorExample 어노테이션 적용

Unagi-zoso commented:
저의 경우 동적으로 예외 메시지를 만드는 커스텀 예외가 있습니다. 이 부분을 ErrorCode 에 정적으로 나타내기가 힘들어
동적인 예외에 한해 기존 방식대로 처리를 했습니다.
그랬더니 어노테이션 방식으로 ErrorExample 을 생성할 때 기존 방식(@Operation(Example ..)) 으로 작성된 부분이 유실되는 경우가
발생했습니다. 이 부분을 해결하기 위해 다음 작업을 했습니다. [77dea44](https://github.com/Weit-2nd/roady-foody-server/pull/32/commits/77dea446fe2bd75417a1669e6f9f2e11f3e2fb34)

작업 내용을 요약하자면 response 에서 기존 내용을 가져와 그곳에 추가하는 식으로 변경했습니다.

배포환경에서 테스트해보니 문제 없었는데 혹시 제가 파악하지 못한 부분이 있다면 피드백 부탁드리겠습니다.

jifrozen0110 reviewed (approved):
고생하셨습니다!
꼼꼼하게 짜여있는 코드인것 같습니다.
코드가 왜 이렇게 짜여졌는지 의도도 잘 설명해주셔서 좋았습니다.

이벤트 동기로 처리하셨는데 저의 개인적인 생각으로는
회원탈퇴를 누르면 클라이언트상에서 로그아웃 처리까지 진행하고 
클라이언트상에 저장한 토큰을 삭제할것같은데
그렇게 되면 실수로라도 accessToken으로 접근이 불가능하지 않을까 싶습니다.

정말 악의적으로 사용자가 accessToken을 따로 저장해두었다
탈퇴하고 다시 접근하려고한다면
이를 막기 위해서는 refreshToken처럼 accessToken도 캐시에 저장해두었다가
삭제하는 방식으로 가야 정확하게 막는것이 아닐까? 라는 생각이 들었습니다.

요약하자면 저의 궁금한점은
비동기에서 동기로 바꿨다고 유효시간이 남은 accessToken의 악의적 접근을 막을 수 있을까? 입니다!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/application/event/AuthEventHandler.kt:
```diff
@@ -0,0 +1,29 @@
+) {
+    @EventListener(AuthLeaveEvent::class)
+    fun handleAuthLeaveEvent(event: AuthLeaveEvent) {
+        val user = userRepository.getByUserId(event.userId)
+        userAgreedTermRepository.deleteAllByUser(user)
+        userRepository.delete(user)
+        val profileImageName = user.profile.profileImageName
+        profileImageName?.let {
+            imageService.remove(profileImageName)
+        }
+        jwtUtil.removeCachedRefreshToken(user.id)
```
요기에 TODO로 다른분들도 회원탈퇴시 해야할 행동 각자 정의 해달라고 적어주시면 좋을것 같아요!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/global/swagger/config/ApiDocsConfig.kt:
```diff
@@ -137,13 +139,23 @@ class ApiDocsConfig {
         responses.addApiResponse(exampleHolder.code.toString(), apiResponse)
     }
+
+    private fun hideParameterByType(
+        operation: Operation,
+        parameterType: String,
+    ) {
+        operation.parameters =
+            operation.parameters?.filterNot { parameter ->
+                parameter.schema?.`$ref`?.endsWith("/$parameterType") == true
+            }
+    }
```
아핫 ㅋㅋㅋ
제의도는 아래 링크처럼 간단하게 static으로 없애는거였어요
https://stackoverflow.com/a/60689297

Unagi-zoso commented:
> 고생하셨습니다! 꼼꼼하게 짜여있는 코드인것 같습니다. 코드가 왜 이렇게 짜여졌는지 의도도 잘 설명해주셔서 좋았습니다.
> 
> 이벤트 동기로 처리하셨는데 저의 개인적인 생각으로는 회원탈퇴를 누르면 클라이언트상에서 로그아웃 처리까지 진행하고 클라이언트상에 저장한 토큰을 삭제할것같은데 그렇게 되면 실수로라도 accessToken으로 접근이 불가능하지 않을까 싶습니다.
> 
> 정말 악의적으로 사용자가 accessToken을 따로 저장해두었다 탈퇴하고 다시 접근하려고한다면 이를 막기 위해서는 refreshToken처럼 accessToken도 캐시에 저장해두었다가 삭제하는 방식으로 가야 정확하게 막는것이 아닐까? 라는 생각이 들었습니다.
> 
> 요약하자면 저의 궁금한점은 비동기에서 동기로 바꿨다고 유효시간이 남은 accessToken의 악의적 접근을 막을 수 있을까? 입니다!

좋은 의견 감사합니다! 일단 저희 인증필터에서 회원이 존재하는지를 검증하기 때문에 회원탈퇴 과정에서 회원만 제대로 지워진다면
유효시간이 남은 accessToken 의 접근 또한 실패할 것이라 생각합니다.

말씀해주신 것처럼 프론트단에서 accessToken 을 관리해주는 만큼 탈취자가 accessToken 을 얻어내는게 개발에 반영될만큼 
현실적인 일인지에 대해서 한 번 더 생각해봐야 할 것 같네요.  그리고 사용자경험 관점에서 현재 탈퇴 기능이  
유저동의약관, 유저 정도의 수준이여서 동기적인 사용이 유효하지 
그동안 작성한 모든 리포트, 모든 댓글을 동기적으로 제거한다 생각하면 비동기적인 처리가 합리적일 수 있다 생각이 드네요.

jifrozen0110 commented:
> 좋은 의견 감사합니다! 일단 저희 인증필터에서 회원이 존재하는지를 검증하기 때문에 회원탈퇴 과정에서 회원만 제대로 지워진다면
> 유효시간이 남은 accessToken 의 접근 또한 실패할 것이라 생각합니다.

아하 이해했습니다!
개인적으로 인간이 accessToken으로 악의적 접근하려고 하기 전애 
서버가 비동기 작업을 완료하지 않을까? 라는 생각도 듭니다!

고생하셨습니다! 👍

Unagi-zoso committed:
ROFO-74 User 파라미터 전역 범위 숨김 방식 변경

Unagi-zoso committed:
ROFO-74 회원탈퇴 기능 TODO 주석 추가

konempty reviewed (approved):
고생하셨습니다 👍 👍 
확실히 코드만 봐도 많은 고민을 하신것 같은데
제가 리뷰를 드린뒤 달리는 답변을 보면서 
다시한번 깊은 고민을 해보신거라는걸 알게되는것 같네요 ㅋㅋㅋ

Unagi-zoso committed:
Merge branch 'develop' into feature/ROFO-74

lin-chae reviewed (approved):
이벤트 방식으로 구현해주신 거 잘 봤습니다. 다른 분들이 질문해주신 거 답변 보면서 고민 많이 하셨을 거 같네요! 고생하셨습니다. 👍
